File: implementations/httpcat-rust/src/cmd.rs
use anyhow::Result;
use clap::Parser;
use reqwest::StatusCode;

use crate::logic::{
    convert_image_to_ascii, fetch_pet_code_image_as_bytes, is_status_code_valid, PetType,
};

#[derive(Debug, Parser)]
pub struct Cmd {
    /// HTTP status code.
    status_code: StatusCode,
    /// Use if you prefer dogs.
    #[arg(long)]
    dog: bool,
    /// Image width.
    #[arg(long, default_value = "100")]
    width: u32,
    /// Image height.
    #[arg(long, default_value = "50")]
    height: u32,
}

pub async fn run() -> Result<()> {
    let args = Cmd::parse();
    let pet_type = if args.dog { PetType::Dog } else { PetType::Cat };

    if !is_status_code_valid(args.status_code) {
        return Err(anyhow::anyhow!(
            "Unsupported status code {}",
            args.status_code
        ));
    }

    let bytes = fetch_pet_code_image_as_bytes(pet_type, args.status_code).await?;
    println!(
        "{}",
        convert_image_to_ascii(bytes, args.width, args.height).await?
    );

    Ok(())
}

File: implementations/httpcat-net/HttpCat/Cmd.cs
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;
using System.Net;
using Spectre.Console.Cli;

namespace HttpCat;

public class Cmd {
    public static int Run(string[] args) {
        var app = new CommandApp<HttpCatCmd>();
        app.Configure(config => {
            config.SetApplicationName("httpcat-net");
        });
        return app.Run(args);
    }
}

internal sealed class HttpCatCmd: Command<HttpCatCmd.Settings>
{
    public sealed class Settings: CommandSettings
    {
        [Description("HTTP status code to fetch.")]
        [CommandArgument(0, "<HTTP status code>")]
        public int StatusCode { get; init; }

        [Description("Use if you prefer dogs.")]
        [CommandOption("--dog")]
        public bool Dog { get; init; }

        [Description("Image width.")]
        [CommandOption("--width")]
        [DefaultValue(100)]
        public int Width { get; init; }
    }

    public override int Execute(CommandContext context, Settings settings)
    {
        var task = Task.Run(async () => await ExecuteAsync(context, settings));
        return task.Result;
    }

    public static async Task<int> ExecuteAsync(CommandContext context, Settings settings) {
        var petType = PetType.Cat;
        if (settings.Dog) {
            petType = PetType.Dog;
        }

        var imageData = await Logic.FetchPetCodeImage(petType, settings.StatusCode);
        // Console.WriteLine(imageData.Length);

        Console.WriteLine(Logic.ConvertImageToAscii(imageData, settings.Width));

        return 0;
    }
}